core/src/commonMain/kotlin/logs/Logging.kt

```diff
@@ -0,0 +1,51 @@
+package com.juul.kable.logs
+
+internal typealias LoggingBuilder = Logging.() -> Unit
+
+public class Logging {
+
+    public enum class Level {
+
+        /** Logs warnings when unexpected failures occur. */
+        Warnings,
+
+        /** Same as [Warnings] plus all events. */
+        Events,
+
+        /** Same as [Events] plus hex representation of I/O data. */
+        Data,
+    }
+
+    public enum class Format {
+
+        /**
+         * Outputs logging in compact format (on a single line per log), for example:
+         *
+         * ```
+         * example message(detail1=value1, detail2=value2, ...)
+         * ```
+         */
+        Compact,
+
+        /**
+         * Outputs logging in multiline format (spanning multiple lines for log details), for example:
+         *
+         * ```
+         * example message
+         *   detail1: value1
+         *   detail2: value2
+         *   ...
+         * ```
+         */
+        Multiline,
+    }
+
+    public fun interface DataProcessor {
+        public fun process(data: ByteArray): String
```

Comment:
Are there ever going to be `ByteArray` details that aren't just the `data` key? If there's potential to have them under multiple keys in the future, I'd love to see a `key` field here even if unused for now. It's a breaking change to retrofit in the future, but only a `_, ` to ignore an extra field.

Reply:
I was thinking a similar thing when I wrote this, that it would be good to know move about the `LogMessage` in the data processor. Knowing this like what characteristic is being written to or read from might drive how the `ByteArray` is processed.

I thought about providing `process` function with the `LogMessage` as well, but it is an internal class (due to its mutability — I wouldn't want to expose it here).

What you're asking specifically is concerning if we later have multiple fields that we want to log that hold data, right?
If, for example, when writing a characteristic it had both the payload it is going to write and potentially some other binary data associated with the write operation?

In which case, would the signature of this function change to?:

```suggestion
        public fun process(key: String, data: ByteArray): String
```

...and then it would be called twice for that write operation? Once for the payload (e.g. `key` of `data`) and again for the additional (we'll call it `extra`) data?

Reply:
> What you're asking specifically is concerning if we later have multiple fields that we want to log that hold data, right? If, for example, when writing a characteristic it had both the payload it is going to write and potentially some other binary data associated with the write operation?

Yes, that's what I was asking about. But your example, exposing the characteristic, is _way more interesting_. A future-proof way to leave room to add fields is to do something like this:

```suggestion
        public fun process(metadata: Metadata, data: ByteArray): String
        public class Metadata internal constructor()
```

Adding fields to a class is a minor/non-breaking change, and we don't have to commit to anything extra right now.